### PR TITLE
Refactor get device by volumeID routines

### PR DESCRIPTION
**Pull request by huww98**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

* Extract logic to gain testability
* Try device link first
* Support NVMe device link
* Return the disk/by-id path if found
* Support for NVMe device in getDeviceBySerial
* Remove search disk by udevadm
* Merge in the logic of GetRootSubDevicePath into adaptDevicePartition
* Add unit test

See commit message for details

#### Which issue(s) this PR fixes:

Fixes the confusing error message when AD controller enabled but disk not found:

Old: NodeStageVolume: ADController Enabled, but device can't be found:d-2ze08oskojgoeh7vceblvolumeID link path "/dev/disk/by-id/virtio-2ze08oskojgoeh7vcebl" not found

New: NodeStageVolume: ADController Enabled, but disk d-2ze08oskojgoeh7vceblvolume can't be found: [get by link "/dev/disk/by-id/virtio-2ze08oskojgoeh7vcebl" failed: file does not exist, get by link "/dev/disk/by-id/nvme-Alibaba_Cloud_Elastic_Block_Storage_2ze08oskojgoeh7vcebl" failed: file does not exist, find by serial: device not found]

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/disk/device_manager.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	// fallback to scan all links
+	if !fileFound {
+		files, err := os.ReadDir(byIDPath)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("read dir %q failed: %w", byIDPath, err))
+		} else {
+			for _, f := range files {
+				if strings.Contains(f.Name(), idSuffix) {
```
Comment (mowangdk) [inline pkg/disk/device_manager.go:178]:
```
if !strings.Contains(f.Name(), idSuffix) {
    continue
}
volumeLinkPath := filepath.Join(byIDPath, f.Name())
...
```
Minimize code nesting levels, make code more readable

File: pkg/disk/device_manager.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+						return volumeLinkPath, nil
+					}
+					errs = append(errs, fmt.Errorf("get by scanned link %q failed: %w", volumeLinkPath, err))
+				}
+			}
+		}
+	}
+
+	// this is danger in Bdf mode
+	if !m.DisableSerial {
```
Comment (mowangdk) [inline pkg/disk/device_manager.go:192]: same as above

File: pkg/disk/cloud.go [hunk 3]
```diff
@@ -233,14 +226,13 @@ func attachDisk(ctx context.Context, tenantUserUID, diskID, nodeID string, isSha
...
-			devicePaths, err = GetDeviceByVolumeID(diskID)
-			deviceName := ""
-			if len(devicePaths) == 0 && bdf != "" {
-				deviceName, err = GetDeviceByBdf(bdf, true)
-			}
-			if err == nil && deviceName != "" {
-				log.Infof("AttachDisk: Successful attach bdf disk %s to node %s device %s by DiskID/Device mapping", diskID, nodeID, deviceName)
-				return deviceName, nil
+			_, err := DefaultDeviceManager.GetRootBlockByVolumeID(diskID)
+			if err != nil && bdf != "" {
```
Comment (mowangdk) [inline pkg/disk/cloud.go:230]: these nest ifs are very confused, try to clarify them.

Comment (huww98) [reply pkg/disk/cloud.go:231]: I think this is out of the scope of this PR. I've reverted some changes here

**Comment by huww98** [comment]
/ok-to-test
@mowangdk I've refactored some nested branches. And some more. PTAL again~

**Comment by mowangdk** [comment]
/lgtm
/approve

Please add partition disk e2e test.

**Pull request merged**